pallets/messages/src/lib.rs

```diff
@@ -185,8 +191,14 @@ impl<T: Config> Pallet<T> {
 
 		'loops: for bid in from..to {
 			let block_number: T::BlockNumber = bid.into();
-			let list = <Messages<T>>::get(block_number, schema_id).into_inner();
+			let retention_block = Self::get_target_retention_block(block_number);
+			let tuple = <Messages<T>>::get(retention_block, schema_id);
+			if tuple.0 != block_number {
+				// this is a left-over block from retention and should be ignored!
+				continue
+			}
 
+			let list = tuple.1.into_inner();
 			let list_size: u32 =
 				list.len().try_into().map_err(|_| Error::<T>::TypeConversionOverflow)?;
```

Comment:
Naming variables with meaningful names instead of tuple, list would help readability